File: php/test/GetChannelInfoEntityTest.php
<?php
declare(strict_types=1);

// GetChannelInfo entity test

require_once __DIR__ . '/../telegramchannel_sdk.php';
require_once __DIR__ . '/Runner.php';

use PHPUnit\Framework\TestCase;
use Voxgig\Struct\Struct as Vs;

class GetChannelInfoEntityTest extends TestCase
{
    public function test_create_instance(): void
    {
        $testsdk = TelegramChannelSDK::test(null, null);
        $ent = $testsdk->GetChannelInfo(null);
        $this->assertNotNull($ent);
    }

    public function test_basic_flow(): void
    {
        $setup = get_channel_info_basic_setup(null);
        // Per-op sdk-test-control.json skip.
        $_live = !empty($setup["live"]);
        foreach (["load"] as $_op) {
            [$_shouldSkip, $_reason] = Runner::is_control_skipped("entityOp", "get_channel_info." . $_op, $_live ? "live" : "unit");
            if ($_shouldSkip) {
                $this->markTestSkipped($_reason ?? "skipped via sdk-test-control.json");
                return;
            }
        }
        // The basic flow consumes synthetic IDs from the fixture. In live mode
        // without an *_ENTID env override, those IDs hit the live API and 4xx.
        if (!empty($setup["synthetic_only"])) {
            $this->markTestSkipped("live entity test uses synthetic IDs from fixture — set TELEGRAMCHANNEL_TEST_GET_CHANNEL_INFO_ENTID JSON to run live");
            return;
        }
        $client = $setup["client"];

        // Bootstrap entity data from existing test data.
        $get_channel_info_ref01_data_raw = Vs::items(Helpers::to_map(
            Vs::getpath($setup["data"], "existing.get_channel_info")));
        $get_channel_info_ref01_data = null;
        if (count($get_channel_info_ref01_data_raw) > 0) {
            $get_channel_info_ref01_data = Helpers::to_map($get_channel_info_ref01_data_raw[0][1]);
        }

        // LOAD
        $get_channel_info_ref01_ent = $client->GetChannelInfo(null);
        $get_channel_info_ref01_match_dt0 = [];
        [$get_channel_info_ref01_data_dt0_loaded, $err] = $get_channel_info_ref01_ent->load($get_channel_info_ref01_match_dt0, null);
        $this->assertNull($err);
        $this->assertNotNull($get_channel_info_ref01_data_dt0_loaded);

    }
}

function get_channel_info_basic_setup($extra)
{
    Runner::load_env_local();

    $entity_data_file = __DIR__ . '/../../.sdk/test/entity/get_channel_info/GetChannelInfoTestData.json';
    $entity_data_source = file_get_contents($entity_data_file);
    $entity_data = json_decode($entity_data_source, true);

    $options = [];
    $options["entity"] = $entity_data["existing"];

    $client = TelegramChannelSDK::test($options, $extra);

    // Generate idmap.
    $idmap = [];
    foreach (["get_channel_info01", "get_channel_info02", "get_channel_info03"] as $k) {
        $idmap[$k] = strtoupper($k);
    }

    // Detect ENTID env override before envOverride consumes it. When live
    // mode is on without a real override, the basic test runs against synthetic
    // IDs from the fixture and 4xx's. Surface this so the test can skip.
    $entid_env_raw = getenv("TELEGRAMCHANNEL_TEST_GET_CHANNEL_INFO_ENTID");
    $idmap_overridden = $entid_env_raw !== false && str_starts_with(trim($entid_env_raw), "{");

    $env = Runner::env_override([
        "TELEGRAMCHANNEL_TEST_GET_CHANNEL_INFO_ENTID" => $idmap,
        "TELEGRAMCHANNEL_TEST_LIVE" => "FALSE",
        "TELEGRAMCHANNEL_TEST_EXPLAIN" => "FALSE",
    ]);

    $idmap_resolved = Helpers::to_map(
        $env["TELEGRAMCHANNEL_TEST_GET_CHANNEL_INFO_ENTID"]);
    if ($idmap_resolved === null) {
        $idmap_resolved = Helpers::to_map($idmap);
    }

    if ($env["TELEGRAMCHANNEL_TEST_LIVE"] === "TRUE") {
        $merged_opts = Vs::merge([
            [
            ],
            $extra ?? [],
        ]);
        $client = new TelegramChannelSDK(Helpers::to_map($merged_opts));
    }

    $live = $env["TELEGRAMCHANNEL_TEST_LIVE"] === "TRUE";
    return [
        "client" => $client,
        "data" => $entity_data,
        "idmap" => $idmap_resolved,
        "env" => $env,
        "explain" => $env["TELEGRAMCHANNEL_TEST_EXPLAIN"] === "TRUE",
        "live" => $live,
        "synthetic_only" => $live && !$idmap_overridden,
        "now" => (int)(microtime(true) * 1000),
    ];
}

File: go/test/get_channel_info_entity_test.go
package sdktest

import (
	"encoding/json"
	"os"
	"path/filepath"
	"runtime"
	"strings"
	"testing"
	"time"

	sdk "github.com/voxgig-sdk/telegram-channel-sdk/go"
	"github.com/voxgig-sdk/telegram-channel-sdk/go/core"

	vs "github.com/voxgig-sdk/telegram-channel-sdk/go/utility/struct"
)

func TestGetChannelInfoEntity(t *testing.T) {
	t.Run("instance", func(t *testing.T) {
		testsdk := sdk.TestSDK(nil, nil)
		ent := testsdk.GetChannelInfo(nil)
		if ent == nil {
			t.Fatal("expected non-nil GetChannelInfoEntity")
		}
	})

	t.Run("basic", func(t *testing.T) {
		setup := get_channel_infoBasicSetup(nil)
		// Per-op sdk-test-control.json skip — basic test exercises a flow
		// with multiple ops; skipping any op skips the whole flow.
		_mode := "unit"
		if setup.live {
			_mode = "live"
		}
		for _, _op := range []string{"load"} {
			if _shouldSkip, _reason := isControlSkipped("entityOp", "get_channel_info." + _op, _mode); _shouldSkip {
				if _reason == "" {
					_reason = "skipped via sdk-test-control.json"
				}
				t.Skip(_reason)
				return
			}
		}
		// The basic flow consumes synthetic IDs from the fixture. In live mode
		// without an *_ENTID env override, those IDs hit the live API and 4xx.
		if setup.syntheticOnly {
			t.Skip("live entity test uses synthetic IDs from fixture — set TELEGRAMCHANNEL_TEST_GET_CHANNEL_INFO_ENTID JSON to run live")
			return
		}
		client := setup.client

		// Bootstrap entity data from existing test data (no create step in flow).
		getChannelInfoRef01DataRaw := vs.Items(core.ToMapAny(vs.GetPath("existing.get_channel_info", setup.data)))
		var getChannelInfoRef01Data map[string]any
		if len(getChannelInfoRef01DataRaw) > 0 {
			getChannelInfoRef01Data = core.ToMapAny(getChannelInfoRef01DataRaw[0][1])
		}
		// Discard guards against Go's unused-var check when the flow's steps
		// happen not to consume the bootstrap data (e.g. list-only flows).
		_ = getChannelInfoRef01Data

		// LOAD
		getChannelInfoRef01Ent := client.GetChannelInfo(nil)
		getChannelInfoRef01MatchDt0 := map[string]any{}
		getChannelInfoRef01DataDt0Loaded, err := getChannelInfoRef01Ent.Load(getChannelInfoRef01MatchDt0, nil)
		if err != nil {
			t.Fatalf("load failed: %v", err)
		}
		if getChannelInfoRef01DataDt0Loaded == nil {
			t.Fatal("expected load result to be non-nil")
		}

	})
}

func get_channel_infoBasicSetup(extra map[string]any) *entityTestSetup {
	loadEnvLocal()

	_, filename, _, _ := runtime.Caller(0)
	dir := filepath.Dir(filename)

	entityDataFile := filepath.Join(dir, "..", "..", ".sdk", "test", "entity", "get_channel_info", "GetChannelInfoTestData.json")

	entityDataSource, err := os.ReadFile(entityDataFile)
	if err != nil {
		panic("failed to read get_channel_info test data: " + err.Error())
	}

	var entityData map[string]any
	if err := json.Unmarshal(entityDataSource, &entityData); err != nil {
		panic("failed to parse get_channel_info test data: " + err.Error())
	}

	options := map[string]any{}
	options["entity"] = entityData["existing"]

	client := sdk.TestSDK(options, extra)

	// Generate idmap via transform, matching TS pattern.
	idmap := vs.Transform(
		[]any{"get_channel_info01", "get_channel_info02", "get_channel_info03"},
		map[string]any{
			"`$PACK`": []any{"", map[string]any{
				"`$KEY`": "`$COPY`",
				"`$VAL`": []any{"`$FORMAT`", "upper", "`$COPY`"},
			}},
		},
	)

	// Detect ENTID env override before envOverride consumes it. When live
	// mode is on without a real override, the basic test runs against synthetic
	// IDs from the fixture and 4xx's. Surface this so the test can skip.
	entidEnvRaw := os.Getenv("TELEGRAMCHANNEL_TEST_GET_CHANNEL_INFO_ENTID")
	idmapOverridden := entidEnvRaw != "" && strings.HasPrefix(strings.TrimSpace(entidEnvRaw), "{")

	env := envOverride(map[string]any{
		"TELEGRAMCHANNEL_TEST_GET_CHANNEL_INFO_ENTID": idmap,
		"TELEGRAMCHANNEL_TEST_LIVE":      "FALSE",
		"TELEGRAMCHANNEL_TEST_EXPLAIN":   "FALSE",
	})

	idmapResolved := core.ToMapAny(env["TELEGRAMCHANNEL_TEST_GET_CHANNEL_INFO_ENTID"])
	if idmapResolved == nil {
		idmapResolved = core.ToMapAny(idmap)
	}

	if env["TELEGRAMCHANNEL_TEST_LIVE"] == "TRUE" {
		mergedOpts := vs.Merge([]any{
			map[string]any{
			},
			extra,
		})
		client = sdk.NewTelegramChannelSDK(core.ToMapAny(mergedOpts))
	}

	live := env["TELEGRAMCHANNEL_TEST_LIVE"] == "TRUE"
	return &entityTestSetup{
		client:        client,
		data:          entityData,
		idmap:         idmapResolved,
		env:           env,
		explain:       env["TELEGRAMCHANNEL_TEST_EXPLAIN"] == "TRUE",
		live:          live,
		syntheticOnly: live && !idmapOverridden,
		now:           time.Now().UnixMilli(),
	}
}
